### DISCUSSION TITLE
stack, vstack, hstack patterns are not working.

### DISCUSSION THREAD

**Discussion by e2goon** [chakra-ui/panda#1034] [Q&A] [answered]
### Description

# stack, vstack, hstack patterns are not working.

![image](https://github.com/chakra-ui/panda/assets/45934/9a070d3b-f09d-43ea-ad4b-89a700b6f12a)

First, I pulled out the stack and then gave it the option `{ gap: '4' }`.

![image](https://github.com/chakra-ui/panda/assets/45934/48d8c74f-c9dd-4d03-9ed8-51421bd7111f)

I checked with the Chrome Console DevTools and the class name came out fine.

![image](https://github.com/chakra-ui/panda/assets/45934/ab80a317-2d87-4450-858e-ed65965d5bc0)

However, it looks like the style code was not added to the layout.css file.

The `css({ display: 'flex', ... )` works fine.

## Use Version

```
- "next": "13.4.10"
- "postcss": "8.4.26"
```

### Link to Reproduction

Not reproducible in Windows environment

https://stackblitz.com/edit/vercel-next-js-n3x2gj?file=app%2Fglobal.css,panda.config.ts,app%2Fpage.tsx,app%2Flayout.tsx,README.md,next.config.js

Still, this is the exact same example as my environment. I need to reproduce this on Windows.

### JS Framework

Next.js (TS) 13 App Dir

### Panda CSS Version

^0.6.0

### Browser

114.0.5735.199

### Operating System

- [ ] macOS
- [x] Windows
- [ ] Linux

**Comment by astahmer** [comment, marked as answer]
the linked minimal repro doesn't look like the screenshots you provided, so it doesn't seem to have an issue right now
![Screenshot 2023-07-15 at 17 24 03](https://github.com/chakra-ui/panda/assets/47224540/a82b1804-ce0f-45bf-99f5-b3edb47d5430)

can you update your repro ?

**Reply by e2goon** [reply]
The Windows environment is where I'm having problems. The stackblitz I provided is fine, because that's not Windows.

**Reply by astahmer** [reply]
I was saying that cause your imports doesn't look the same in the screenshots you sent (`~/styled-system/css`, probably using custom tsconfig paths) VS in the stackblitz (`../styled-system/css`)

**Reply by e2goon** [reply]
Sorry about that. Batch changing the path to `outdir: 'styled-system'` inside the `panda.config.ts` file, and changing it to a relative path instead of an absolute path fixed it.

I set the directory `outdir: 'src/styled-system'` inside the panda.config.ts file and wanted to use the path `~/styled-system/...` to look nice in the tsconfig.json. But this was the wrong choice.

Thank you 🙏

**Reply by osdiab** [reply]
#868 tracks that and #1039 should fix it

**Comment by Sadeghi1017** [comment]
**vdhni**

**Comment by wmertens** [comment]
I'm also not getting any output from patterns. Is there supposed to be CSS generated for `hstack` or `circle`?

Here's the repro with Next: https://stackblitz.com/edit/vercel-next-js-z7gzeg?file=panda.config.ts,src%2Fapp%2Fpage.tsx,src%2Fapp%2Flayout.tsx
And this one's with Qwik, it loads faster: https://stackblitz.com/edit/qwik-starter-asn8v4?file=panda.config.ts,src%2Froutes%2Findex.tsx,src%2Fstyled-system%2Fdebug%2Fsrc__routes__index.css

**Reply by astahmer** [reply]
Hey, it's already fixed but not yet merged & published on npm
it'll be available once [this PR](https://github.com/chakra-ui/panda/pull/1055) gets merged
and [this](https://github.com/chakra-ui/panda/commit/be0ad578156d1bed81ab26be770569674b56fb3d) was the PR that fixed this issue

**Comment by osdiab** [comment]
You can use it for now by installing ***@***.******@***.***`

Omar


On Tue, Jul 25, 2023 at 5:36 AM, Alexandre Stahmer ***@***.***

**Comment by wmertens** [comment]
Thanks! I made a new blitz, this uses Manu's macro plugin for all styled-system imports, and the resulting client code does not contain any imports from styled-system :partying_face: 

https://stackblitz.com/edit/qwik-starter-vuepi6?file=src%2Froutes%2Findex.tsx,vite.config.ts

BTW:
- the Qwik install documentation should just say "run `npx qwik add pandacss`", it adds a bunch of other things like ignores, a demo path, and the correct script name for the build
- how does pandacss work to inject the CSS in dev mode? I can't figure it out :smiling_face_with_tear:

**Reply by astahmer** [reply]
this looks amazing ! I was curious to see how it's implemented but can't find the source for this macro plugin, is it private ?

> how does pandacss work to inject the CSS in dev mode? I can't figure it out 🥲

I guess you're using the postcss plugin ? if that's the case, postcss is the one doing the heavy work here
otherwise, when using the Panda CLI only (with watch flag), that's up to you to import the generated `styles.css` and there is a merge happening on this file between old (on disk) / new version (re-generated on file change)

**Reply by wmertens** [reply]
The plugin is at https://github.com/manucorporat/vite-plugin-macro

How does postcss get the CSS for the classes?

**Reply by astahmer** [reply]
thanks !

we have a `Builder` class that gets called on each file change from the postcss plugin
https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/postcss/src/index.ts#L26-L27

then the builder calls the main entrypoint (`loadConfigAndCreateContext`) to resolve the panda.config and prepare everything needed from there
https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/node/src/builder.ts#L136

which also emit the `styled-system` artifacts (runtime .js / .d.ts types) on the first setup call, for convenience so you don't have ot run `panda codegen`
https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/node/src/builder.ts#L140

after that, we extract styles from each file included using globs
https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/node/src/builder.ts#L172-L195

finally, we write the css output into the root `rule` provided by postcss now
https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/postcss/src/index.ts#L37

I also made this visual (it's for the CLI implementation tho, not postcss integration), maybe it helps a bit
https://stately.ai/registry/editor/d30cb4f3-2279-41a6-ad95-13e72ccfe884?machineId=0223298d-5951-46a5-82d5-ef379c701d07&mode=Simulate

**Reply by wmertens** [reply]
Thank you, that's great! So if I understand correctly, Vite has postcss support built-in, and the plugin maintains the styles that should be added to the css. I imagine the `@layer ...;` statement gets replaced?

How do you generate the CSS? For each source file, you generate the AST and run the utility functions, which generate the CSS?

I'm wondering if that step can be sped up, since the macro plugin runs those too. Perhaps the utility functions could import a virtual module from the postcss plugin to register CSS during the macro expansion?

Also
> which also emit the styled-system artifacts (runtime .js / .d.ts types) on the first setup call, for convenience so you don't have ot run panda codegen

This doesn't seem to be happening (in time?), I had to configure the stackblitz to run it before the npm start.

**Reply by astahmer** [reply]
>  So if I understand correctly, Vite has postcss support built-in, and the plugin maintains the styles that should be added to the css.

yes

> I imagine the @layer ...; statement gets replaced?

no, it's now [native to CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer) so we can keep it
we do check [if the user has provided the layers we use](https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/node/src/builder.ts#L212) tho

> How do you generate the CSS? For each source file, you generate the AST and run the utility functions, which generate the CSS?

also using postcss, this time internally (meaning it works that way as well for the CLI integration)

the process is like: 
- extract from AST which styles were used 
https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/parser/src/parser.ts#L233-L400
- process those style objects depending on their kind (css/cva/recipe/pattern)
https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/generator/src/artifacts/css/parser-css.ts#L12-L79

finally, the internal process function that converts from a style object to the equivalent CSS is mostly here
https://github.com/chakra-ui/panda/blob/e49e67da40183879dbf1721da8d05f8be95597fc/packages/core/src/atomic-rule.ts#L48-L112

> I'm wondering if that step can be sped up

isn't it fast enough ? you can run the panda command with the env var `PANDA_DEBUG=* pnpm panda` to see the timings, you could use the `panda analyze --json report.json` command to get a JSON output with those timings, the unit is in milliseconds

```json
{
  "duration": {
    "extractTimeByFiles": {
      "/Users/astahmer/dev/open-source/panda/sandbox/vite-ts/src/App.tsx": 9.364959001541138,
      "/Users/astahmer/dev/open-source/panda/sandbox/vite-ts/src/Badge.tsx": 0.6717090606689453,
      "/Users/astahmer/dev/open-source/panda/sandbox/vite-ts/src/Button.tsx": 1.7605409622192383,
      "/Users/astahmer/dev/open-source/panda/sandbox/vite-ts/src/Card.tsx": 0.9269590377807617,
      "/Users/astahmer/dev/open-source/panda/sandbox/vite-ts/src/main.tsx": 0.30408310890197754
    },
    "extractTotal": 13.02825117111206,
    "classify": 2.472208023071289
  },
```

> Perhaps the utility functions could import a virtual module from the postcss plugin to register CSS during the macro expansion?

sorry, I'm not sure what this means 😭  can you show me an example ?

> This doesn't seem to be happening (in time?), I had to configure the stackblitz to run it before the npm start.

i'll check that asap

**Comment by wmertens** [comment]
> isn't it fast enough ?

Well if I see an optimization opportunity I can't not think about it :-) Suppose you have a big project with 100 big source files, 10ms per file would mean an extra second of build time.
Ok, the extract time seems not too bad. :)

> > Perhaps the utility functions could import a virtual module from the postcss plugin to register CSS during the macro expansion?
> sorry, I'm not sure what this means 😭 can you show me an example ?

Hmm, so in vite each import goes through resolveId/load/transform. Vite plugins can intercept these steps to do their thing. The styled-system utils could import `"\0panda-register"` which would be intercepted by the pandacss plugin to somehow store the required CSS during the macro run. But that's probably way too many moving parts to keep working reliably.

Thanks for all the explanations!

**Comment by masylum** [comment]
I still see this issue. If I use the JSX components it does work, but using the pattern functions it does not.

**Reply by strvbkzhn** [reply]
I've also spend some time with this today
```
<Stack direction="vertical"  // doesn't works
<Stack direction="column"    // works fine
```
however in docs said that
![image](https://github.com/chakra-ui/panda/assets/42301364/4f0e4dba-ff88-40d1-9f2f-dd34cbd73371)

**Comment by cs-clarence** [comment]
issue still persists, no problem in linux but windows has problems
